packages/transaction/src/tx-builder.ts

```diff
@@ -299,14 +301,59 @@ export class TxBuilder {
       if (spending.eth().gt(0)) {
         spendings.push(spending)
       }
+      if (spending.asset.erc20Amount.gt('0')) {
+        utxosWithERC20.push(spending)
+      }
+      if (spending.asset.nft.gt('0')) {
+        utxosWithNft.push(spending)
+      }
     }
 
     // Calculate ETH change
     assert(spendingAmount().eth.gte(getRequiredETH()), 'not enough eth')
     const changeETH = spendingAmount().eth.sub(getRequiredETH())
     const finalFee = getTxFee()
     if (!changeETH.isZero()) {
-      changes.push(Utxo.newEtherNote({ eth: changeETH, owner: this.changeTo }))
+      changes.push(
+        Utxo.newEtherNote({
+          eth: changeETH,
+          owner: this.changeTo,
+        }),
+      )
+    }
+
+    // check if ETH spending utxos includes ERC20 or nft
+    if (utxosWithERC20.length > 0) {
```

Comment:
Here is an example for this fix. Let's say Alice has two utxos, 
1. 20 ZRC 
2. 300 ZRC and 1.1 ETH
and gas fee is 0.1 ETH

She wants to withdraw 20(or less) ZRC. In our original scenario, her remaining asset will become 1.1 ETH - 0.1ETH = 1 ETH only, no ZRC anymore (it's bcs #L317 only calculate ETH, 300 ZRC in the same utxo wasn't counted in).

Reply:
Following task created, [#409](https://github.com/zkopru-network/zkopru/issues/409)

Reply:
Great! I wrote [test code](https://gist.github.com/sifnoc/9b00e7d692674652bf798546bfa6c0ca) for this scenario, it worked well as expected.
I think we should add more test cases for `tx-builder` with a combination of different types of utxo

Reply:
agree! #409 is for that purpose to add more test cases.